primer/src/Primer/EvalFullInterp.hs

```diff
@@ -0,0 +1,437 @@
+  -- Consider  @(λx.(λy.x) : A -> B -> A) s t@ we will
+  --   interp' @λy.x@ in context where @x:->t@, and this is the only time we have @x@ in the context!
+  LAM _ v t -> let v' = freshLike v env in LAM () v' $ interp' tydefs (extendTyEnv v (TVar () v') env) Chk t
+  Var _ (LocalVarRef v) -> upsilon dir $ envTm.env ! Right v -- recall the environment is in normal form
+  Var _ (GlobalVarRef v) -> upsilon dir $ envTm.env ! Left v -- recall the environment is in normal form
+  Let _ v e b -> interp' tydefs (extendTmEnv (Right v) (interp' tydefs env Syn e) env) dir b
+  LetType _ v t b -> interp' tydefs (extendTyEnv v (interpTy envTy t) env) dir b
+  -- this interpretation af letrec can easily cause deadlocked or infinite-size programs
+  -- and can't be detected but has advantage of being lazy and having sharing
+  Letrec _ v e t b ->
+    let e' = interp' tydefs env' Chk e
+        env' =
+          extendTmEnvWithFVs
+            (Right v)
+            (Ann () e' $ interpTy envTy t)
+            (Set.delete (unLocalName v) $ freeVars (Ann () e t))
+            env
+     in interp' tydefs env' dir b
+  -- In step interpreter, case which does not discriminate is lazy. Same here for consistency.
+  Case _ _ [] (CaseFallback e) -> interp' tydefs env Chk e
+  Case _ e brs fb ->
+    -- recall @case C as : T A of ... ; C xs -> e ; ...@ steps to
+    --  @let xs=as:(lettype p=A in S) in e@ for @data T p = C S@
+    case interp' tydefs env Syn e of
+      Ann _ (Con _ c as) (decomposeTAppCon -> Just (tycon, tyargs))
+        | Just (CaseBranch _ xs t) <- find ((PatCon c ==) . caseBranchName) brs ->
+            let envTy' = extendTysEnv' (tyParamEnvExt tycon tyargs) envTy
+             in interp'
+                  tydefs
+                  ( extendTmsEnv
+                      ( zip (Right . bindName <$> xs)
+                          $ zipWith (\a argTy -> Ann () a $ interpTy envTy' argTy) as
+                          $ ctorArgTys tycon c
+                      )
+                      env
+                  )
+                  Chk
+                  t
+        | CaseFallback t <- fb -> interp' tydefs env Chk t
+        | otherwise -> throw $ NoBranch (Left c) $ caseBranchName <$> brs
+      Ann _ (PrimCon _ c) (TCon _ ((== primConName c) -> True))
+        | Just (CaseBranch _ [] t) <- find ((PatPrim c ==) . caseBranchName) brs -> interp' tydefs env Chk t
+        | CaseFallback t <- fb -> interp' tydefs env Chk t
+        | otherwise -> throw $ NoBranch (Right c) $ caseBranchName <$> brs
+      -- literals (primitive constructors) are actually synthesisable, so may come
+      -- without annotations
+      PrimCon _ c
+        | Just (CaseBranch _ [] t) <- find ((PatPrim c ==) . caseBranchName) brs -> interp' tydefs env Chk t
+        | CaseFallback t <- fb -> interp' tydefs env Chk t
+        | otherwise -> throw $ NoBranch (Right c) $ caseBranchName <$> brs
```

Comment:
This "duplication" is a bit of a worry. It is reflecting the same in the stepwise evaluator, but we should perhaps review the calculus to see if it is really needed.

Reply:
See https://github.com/hackworthltd/primer/issues/1237